### ISSUE TITLE
Mount randomly drops

### ISSUE THREAD

**Issue by GhyslainBruno** [astrada/google-drive-ocamlfuse#592]
Hi all, 

I'm reopening a closed issue, which, in my opinion, shouldn't be ([#281](https://github.com/astrada/google-drive-ocamlfuse/issues/281))

I don't know if this issue is solved with the latest version for you guys, but for me I had this weird error, very randomly, and sometimes after few days : 

```
ls: cannot access 'google_drive': Transport endpoint is not connected
```

*My mount point folder is "google_drive"* 

So I decided to write a little bash script as a workaround, which runs every minute in a crontab, and if my mount is not available, use fusermount to unmount the mount point, and do it again. 

**Since this, it works all the time** 
(fyi, I use google-drive-ocamlfuse for/with Plex, so it is really important to keep files available all the time).

My crontab : 
````
@reboot sleep 60 &&  /home/USERNAME/bin/automount_google_drive.sh
* * * * * /home/USERNAME/bin/check_google_drive_mounting.sh
````

check_google_drive_mounting.sh file : 
```
#!/bin/bash
ls /home/USERNAME/google_drive
if [ $? -eq 0 ]; then
    echo OK
else
    echo FAIL - mounting again
    sudo fusermount -zu /home/USERNAME/google_drive
    /home/USERNAME/bin/automount_google_drive.sh
fi
```

automount_google_drive.sh file : 
````
#!/bin/bash

sudo google-drive-ocamlfuse -o allow_other -label google_drive_mount /home/USERNAME/google_drive/
````

PS : 
* [Use sudo without typing password](https://phpraxis.wordpress.com/2016/09/27/enable-sudo-without-password-in-ubuntudebian/)
* Platform : Ubuntu 18.04.3 LTS on an Intel Nuc

_Originally posted by @GhyslainBruno in https://github.com/astrada/google-drive-ocamlfuse/issues/281#issuecomment-553480626_

**Comment by brancomat** [comment]
What version are you using?

I have a similar issue (that has become less frequent after 0.7.12), the problem is that to address it one should enable debug mode and collect  `gdfuse.log` and `curl.log` files, and they get huge after a few days.

**Comment by GhyslainBruno** [comment]
I'm using the 0.7.13 
I didn't update to 0.7.14 since I use google-drive-ocamlfuse only in a detach server (using ssh), so I don't need this update for now. 

Hum, I could try this as a first move you're right. The thing is it's pretty rare, sometimes it can happen like once in a week... 

Just to be efficient, do you know how to enable gdfuse and culr logs ? 

Thanks for your answer anyway

**Comment by brancomat** [comment]
> Just to be efficient, do you know how to enable gdfuse and culr logs ?

by adding the `-debug` flag to the `google-drive-ocamlfuse` launch (see https://github.com/astrada/google-drive-ocamlfuse/wiki/Troubleshooting)

If you want to keep your `check_google_drive_mounting.sh` active, you could modify it to copy the log files somewhere else (and maybe timestamp the names) before launching again the mount, something like:

```
else
    echo FAIL - mounting again
    mv ~/.gdfuse/google_drive_mount/gdfuse.log /some/path/$(date +"%Y%m%d%H%M")_gdfuse.log
    mv ~/.gdfuse/google_drive_mount/curl.log /some/path/$(date +"%Y%m%d%H%M")_curl.log
    sudo fusermount -u /home/USERNAME/google_drive
    /home/USERNAME/bin/automount_google_drive.sh
fi
```

some caveats:
- adding the `-debug` flag keeps the process in foreground so the `google-drive-ocamlfuse` command will NOT exit
- if I remember correctly when using debug with labels the log goes in the `~/.gdfuse/$label` folder (and not in the `~/.gdfuse/default` folder as specified in the wiki) but I'm not completely sure.

**Comment by GhyslainBruno** [comment]
Thanks a lot, I'll make these changes asap. 

I'll let you know as soon as I get some info about this ;-). 

@brancomat how do you deal with these annoying random drops on your side ?

**Comment by brancomat** [comment]
> @brancomat how do you deal with these annoying random drops on your side ?

basically the same way as you do, the only difference is a `timeout` before the `ls`:
```
(...)
for i in label1 label2 label3; do
    timeout -v 20 ls /mnt/gdrive/${i}/ 1>/dev/null 2>>test_gdrive_mountpoint.log || { do_things ; }
(...)
```

also: since a couple of times (with older versions) the mountpoint got stuck and the `fusermount` command didn't work, in one of the servers there's an additional check that triggers a reboot if it can't unmount/remount the path (I'm not really proud of that solution).

**Comment by astrada** [comment]
If `fusermount` fails, you should kill the process with `killall -9 google-drive-ocamlfuse`, then you can unmount and remount.

**Comment by GhyslainBruno** [comment]
@brancomat thanks for your infos. 

@astrada I might do this in my scripts you're right, even if the best would be not having to do such a thing :-).

Today I experienced a ```Transport endpoint is not connected``` error, but my scripts didn't work because of a ```ressource busy``` error (sadly I didn't log anything before that :-/). 

I then modified my check_google_drive_mounting.sh file that way : 

- Old : 
```
...
sudo fusermount -u /home/USERNAME/google_drive
...
```

- New : 
```
...
sudo fusermount -zu /home/USERNAME/google_drive
...
```

_[Here is why](https://superuser.com/questions/207269/sshfs-device-is-busy) I added the -z option. I works like a charm,  but I don't really think it's a proper option though._

I then rerun the whole thing in debug mode, with some logging as you pointed @brancomat, we'll see what will happen then. 

Do you guys have an idea of what could cause this random drops ? 
It's really a showstopper to get a proper Plex/Google drive unlimited infrastructure working :-/ 

Anyway, thanks for this discussion.

**Comment by Kibouo** [comment]
This happens to me when my device goes to hibernation mode. My settings are default. The dir is basically in constant use.

[gdfuse.log](https://github.com/astrada/google-drive-ocamlfuse/files/3912526/gdfuse.log)

**Comment by GhyslainBruno** [comment]
@Kibouo thanks for pointing this out, I actually didn't think about that (hibernation mode). 

About the gdfuse.log of 1Gb size I don't know if it's useful (nevertheless, @astrada should have a better answer than me about that).

From my side, I have several log files and their size is > 8Go (actually, **I had to stop keeping the logs** and being in debug, because logging was fulfilling my disk space... )

Even it's not really helpful for now, I wanted to share this with you guys :-/

**Comment by AnselmC** [comment]
Having a similar issue. Drive only stays mounted for a very short period which makes Colab notebooks unusable. 
Not sure is this is due to api restrictions from google. Can't decipher any error in the logs.
Last lines read:
```bash
[557.200617] TID=907: releasedir / O_RDONLY
[557.201287] TID=908: getattr /Colab Notebooks/<ipython-input-10-54178a50f518>
[557.201307] TID=908: BEGIN: Getting metadata from context
[557.201313] TID=908: END: Getting metadata: Valid
[557.201323] TID=908: BEGIN: Loading resource /Colab Notebooks/<ipython-input-10-54178a50f518> (trashed=false) from db
[557.201338] TID=908: END: Loading resource /Colab Notebooks/<ipython-input-10-54178a50f518> (trashed=false) from db: Not found
[557.201344] TID=908: BEGIN: Loading resource /Colab Notebooks (trashed=false) from db
[557.201354] TID=908: END: Loading resource /Colab Notebooks (trashed=false) from db: Found (id=9, state=Synchronized)
[557.201369] TID=908: File not found: stat /Colab Notebooks/<ipython-input-10-54178a50f518>
[571.025141] TID=2: Flushing DB...done
```

Need to reconnect to the drive every time this happens, every few minutes...
Supposedly I could use a cron script as well

**Comment by GhyslainBruno** [comment]
Having this issue every few minutes is weird. 
For me it was (I mean, it is, but I don't really see that anymore since I use the bash scripts I shared here) only every few days, maybe once or twice a week. 
But I assume it depends on the use you have. 
Anyway, maybe we should dig into the source code to help @astrada about this thing

**Comment by GhyslainBruno** [comment]
Hum. I actually don't even know what language is used... 

I'm just a simple JS developer, so I'm afraid I can't really help on that issue personally :-/ 

I hope someone will find time for that anyway. 

Thanks for your help guys

**Comment by astrada** [comment]
I think I found the problem. Now if you mount the fs offline, you should be able to access your Drive when you get back online. The fix is in version 0.7.15 that I'm publishing on the beta ppa and on opam.
